core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,198 @@
+
+class UpdateSnapshotReferencesOperation implements PendingUpdate<Map<String, SnapshotRef>> {
+
+  private final TableOperations ops;
+  private final Map<String, SnapshotRef> updatedRefs;
+  private TableMetadata base;
+
+  UpdateSnapshotReferencesOperation(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.updatedRefs = Maps.newHashMap(base.refs());
+  }
+
+  @Override
+  public Map<String, SnapshotRef> apply() {
+    return refDiff(base.refs(), updatedRefs);
+  }
+
+  @Override
+  public void commit() {
+    TableMetadata updated = internalApply();
+    ops.commit(base, updated);
+  }
+
+  public UpdateSnapshotReferencesOperation createBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef branch = SnapshotRef.branchBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, branch);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation createTag(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef tag = SnapshotRef.tagBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, tag);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeBranch(String name) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such branch with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    updatedRefs.remove(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeTag(String name) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such tag with name %s", name);
+    ValidationException.check(ref.isTag(), "Ref with name %s is a branch not a tag", name);
+    updatedRefs.remove(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation renameBranch(String name, String newName) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    Preconditions.checkNotNull(newName, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such ref with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    updatedRefs.put(newName, ref);
```

Comment:
Looks like this needs to validate the return value from `put` or else this will replace branches.